agent/agent.go

```diff
@@ -242,28 +242,43 @@ func (a *Agent) flush() {
 			}
 		}(o)
 	}
-
 	wg.Wait()
 }
 
 // flusher monitors the metrics input channel and flushes on the minimum interval
 func (a *Agent) flusher(shutdown chan struct{}, metricC chan telegraf.Metric) error {
+	var metricStream chan telegraf.Metric
+	stopFilter := make(chan struct{}, len(a.Config.Filters))
 	// Inelegant, but this sleep is to allow the Gather threads to run, so that
 	// the flusher will flush after metrics are collected.
 	time.Sleep(time.Millisecond * 200)
-
 	ticker := time.NewTicker(a.Config.Agent.FlushInterval.Duration)
+	metricStream = metricC
+	if len(a.Config.Filters) != 0 {
+		for _, name := range a.Config.FiltersOrder {
+			filter := a.Config.Filters[name]
+			log.Printf("Filter %s is enabled", name)
+			metricStream = filter.Pipe(metricStream)
+			go func(f telegraf.Filter) {
```

Comment:
Filter suppose to work in is own context like other plugins it may have timer inside and i don't want to block the agent code from collecting metrics or flushing them. consider filter as pipe with input and output (black box). Histogram filter for example has timer for flushing metrics on specific intervals